src/filters/matches.rs

```diff
@@ -0,0 +1,159 @@
+crate::include_proto!("quilkin.extensions.filters.matches.v1alpha1");
+
+use self::quilkin::extensions::filters::matches::v1alpha1::Matches as ProtoConfig;
+use crate::filters::{prelude::*, registry::FilterRegistry};
+
+pub const NAME: &str = "quilkin.extensions.filters.matches.v1alpha1.Matches";
+
+/// Creates a new factory for generating capture filters.
+pub fn factory() -> DynFilterFactory {
+    Box::from(MatchesFactory::new())
+}
+
+struct Matches {
+    config: Config,
+    metrics_registry: prometheus::Registry,
+    filter_registry: FilterRegistry,
+}
+
+impl Matches {
+    fn new(
+        config: Config,
+        filter_registry: FilterRegistry,
+        metrics_registry: prometheus::Registry,
+    ) -> Self {
+        Self {
+            config,
+            filter_registry,
+            metrics_registry,
+        }
+    }
+}
+
+/// The actual behaviour of the matches filter, the only thing that changes
+/// between read and write is which config and direction is called.
+macro_rules! matches_filter {
+    ($this:ident, $context:ident, $direction:ident, $config:ident) => {{
+        let ctx = $context;
+        let config = match $this.config.$config.as_ref() {
+            Some(value) => value,
+            None => return Some(ctx.into()),
+        };
+
+        for branch in &config.branches {
```

Comment:
Let's increment counter metrics here for when we find a match vs not, for visibility

Reply:
I'm going to add metrics in a follow-up PR since this is already pretty big.